File: 402. Remove K Digits/solution.py
class Solution(object):
    def removeKdigits(self, num, k):
        """
        :type num: str
        :type k: int
        :rtype: str
        """
        # Handle the edge cases
        if len(num) == k:
            return "0"
        
        if len(num) - num.count("0") <= k:
            return "0"
        
        # Remove the first number from left until the order is increasing
        for K in range(k):
            maxi = len(num) - 1
            for i in range(1, len(num)):
                if num[i-1] > num[i]:
                    maxi = i-1
                    print(maxi)
                    break
            
            num = num[:maxi] + num[maxi+1:]    
                
        # remove leading zeroes if any
        return num.lstrip("0")

File: 402. Remove K Digits/solution.java
class Solution {
    public String removeKdigits(String num, int k) {
        // edge cases
        if (num.length() <= k) {
            return new String("0");
        }
        
        // convert to a list
        List<Integer> list = new ArrayList<Integer>(num.length());
        for (int i = 0; i < num.length(); i++) {
            list.add(Character.getNumericValue(num.charAt(i)));
        }
        
        int i = 1;
        while (i < list.size() && k > 0) {
            if (i > 0 && list.get(i-1) > list.get(i)) {
                list.remove(i-1);
                k--;
                i--;
            }
            else  
                i++;
        }
        
        // remove any remaining numbers from left
        while (k > 0) {
            list.remove(list.size()-1);
            k--;
        }
        
        // check for any leading 0s
        i = 0;
        while (i < list.size()) {
            if (i == 0 && list.get(i) == 0) {
                list.remove(i);
                i--;
            }
            i++;
        }
        if (list.size() == 0) {
            return new String("0");
        } 
        
        // convert list to String
        StringBuilder sb = new StringBuilder();
        for (i = 0; i < list.size(); i++) {
            sb.append(String.valueOf(list.get(i)));
        }
        return sb.toString();
    }
}
